ch_pipeline/hfb/flagging.py

```diff
@@ -162,170 +161,197 @@ def process(self, stream):
-        n0, _, n2 = np.meshgrid(
-            np.arange(nfreq), np.arange(nbeam_ns), np.arange(ntime), indexing="ij"
+        angles = v.get_beam_positions(rfimaskbitmap.beam_ns[:], rfimaskbitmap.freq[:])
+        el = np.sin(np.deg2rad(angles.T[1]))  # shape (freq, beam_ns)
+
+        # Create new elevation axis, restricting to overlapping region with original beams
+        to_ax = np.linspace(-1, 1, self.npix)
+        start = _search_nearest(to_ax, np.min(el))
+        end = _search_nearest(to_ax, np.max(el))
+        valid_range = slice(start, end + 1)
+        new_ax = to_ax[valid_range]
+
+        # Create output container with the new axes
+        out = LocalizedRFIMask(
+            freq=rfimaskbitmap.freq[:], el=new_ax, time=rfimaskbitmap.time[:]
         )
-        el_closest_indices = el_closest_indices - np.min(el_closest_indices)
-
-        # Broadcast indices along the last axis
-        index_tuple = (n0, el_closest_indices[sf:ef, :, None], n2)
-
-        # Broadcast the mask data
-        mask_adj = np.full((nfreq, nel, ntime), False)
-        mask_adj[index_tuple] = mask
-
-        # Falg before and after a given true value for conservative flagging
-        for repeat in range(self.spread_size):
-            mask_adj[:, :-1, :] |= mask_adj[:, 1:, :]
-            mask_adj[:, 1:, :] |= mask_adj[:, :-1, :]
-            mask_adj[:, :-2, :] |= mask_adj[:, 2:, :]
-            mask_adj[:, 2:, :] |= mask_adj[:, :-2, :]
-
-        # Create container to hold output
-        out = LocalizedRFIMask(freq=freq, el=new_el, time=time)
-
-        # Save the adjusted mask to output container
-        out.mask[:] = mask_adj
-
-        # If the input container has the frac_rfi dataset
-        if rfimask.datasets.get("frac_rfi", None) is not None:
-            frac_rfi = rfimask.frac_rfi[:]
-            frac_rfi_adj = np.full((nfreq, nel, ntime), 0.0)
-            frac_rfi_adj[index_tuple] = frac_rfi
+        if self.keep_frac_rfi:
+            out.add_dataset("frac_rfi")
 
-            for repeat in range(self.spread_size):
-                frac_rfi_adj[:, :-1, :] = np.maximum(
-                    frac_rfi_adj[:, :-1, :], frac_rfi_adj[:, 1:, :]
+        # Determine local frequency slice for this MPI rank
+        freq_ax = list(rfimaskbitmap.subfreq_rfi.attrs["axis"]).index("freq")
+        nfreq_local = rfimaskbitmap.subfreq_rfi[:].local_array.shape[freq_ax]
+        max_freq = self.comm.allreduce(nfreq_local, op=MPI.MAX)
+        sf = rfimaskbitmap.subfreq_rfi.local_offset[freq_ax]
+        ef = sf + max_freq + 1
+
+        # Process each frequency separately since beam-to-el mapping varies with freq
+        for f in range(sf, ef):
+
+            # To avoid issues occurring when the last iteration of frequency-distributed data processing
+            # is pending, particularly to handle the case of an empty slice at the end.
+            i = f
+            if (i - sf) >= nfreq_local:
+                i = sf
```

Comment:
I would check if this is needed now - if not, I would remove it to make it clearer for a reader.

I _think_ that the only blocking MPI call (which was causing this to hang) was in the creation of a new container on each iteration. Since that isn't happening anymore, each MPI process should be ok with a different number of frequencies.